Quarkus fails to get GraalVM version from GraalVM CE builds based on OpenJDK instead of LabsJDK

### Describe the bug

Using a GraalVM build based on OpenJDK instead of LabsJDK results in build failures of native executables.

### Expected behavior

The build should succeed if it succeeds with a LabsJDK based build of GraalVM.

### Actual behavior

The build fails with:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project quarkus-integration-test-picocli-native: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.RuntimeException: Failed to get GraalVM version
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildRunner.getGraalVMVersion(NativeImageBuildRunner.java:41)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:236)
[ERROR] 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:854)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[ERROR] Caused by: java.lang.NumberFormatException: For input string: "null"
[ERROR] 	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
[ERROR] 	at java.base/java.lang.Integer.parseInt(Integer.java:662)
[ERROR] 	at java.base/java.lang.Integer.parseInt(Integer.java:778)
[ERROR] 	at java.base/java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:214)
[ERROR] 	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
[ERROR] 	at java.base/java.util.stream.IntPipeline.toArray(IntPipeline.java:562)
[ERROR] 	at io.quarkus.deployment.pkg.steps.GraalVM$Version.breakdownVersion(GraalVM.java:240)
[ERROR] 	at io.quarkus.deployment.pkg.steps.GraalVM$Version.<init>(GraalVM.java:229)
[ERROR] 	at io.quarkus.deployment.pkg.steps.GraalVM$VersionParseHelper.parse(GraalVM.java:89)
[ERROR] 	at io.quarkus.deployment.pkg.steps.GraalVM$Version.of(GraalVM.java:306)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildRunner.getGraalVMVersion(NativeImageBuildRunner.java:38)
[ERROR] 	... 10 more
```

### How to Reproduce?

1. Build GraalVM from source using OpenJDK as the base JDK (or grab a build from the artifacts of https://github.com/graalvm/graalvm-community-jdk21u/actions/workflows/quarkus.yml)
2. Set `GRAALVM_HOME` to point to the freshly built GraalVM
3. Try to generate a native executable

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Mandrel or GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

This is happening because Quarkus currently has not way to infer the GraalVM version from the `--version` output, and fails to handle the failure as well.

The `--version` output that can't be processed by Quarkus is:

```
native-image 21.0.5-beta 2024-10-15
GraalVM Runtime Environment GraalVM CE 21.0.5-dev.beta+4.1 (build 21.0.5-beta+4-ea)
Substrate VM GraalVM CE 21.0.5-dev.beta+4.1 (build 21.0.5-beta+4-ea, serial gc)
```

--- a/core/deployment/src/test/java/io/quarkus/deployment/pkg/steps/GraalVMTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/pkg/steps/GraalVMTest.java
@@ -177,6 +177,20 @@ public class GraalVMTest {
     }
 
     @Test
+    public void testGraalVM23_1CommunityVersionParser() {
+        final Version version = Version.of(Stream.of(("native-image 21.0.5-beta 2024-10-15\n"
+                + "GraalVM Runtime Environment GraalVM CE 21.0.5-dev.beta+3.1 (build 21.0.5-beta+3-ea)\n"
+                + "Substrate VM GraalVM CE 21.0.5-dev.beta+3.1 (build 21.0.5-beta+3-ea, serial gc)")
+                .split("\\n")));
+        assertThat(version.toString().contains(GRAALVM.name()));
+        assertThat(version.getVersionAsString()).isEqualTo("23.1-dev");
+        assertThat(version.javaVersion.toString()).isEqualTo("21.0.5-beta+3-ea");
+        assertThat(version.javaVersion.feature()).isEqualTo(21);
+        assertThat(version.javaVersion.interim()).isEqualTo(0);
+        assertThat(version.javaVersion.update()).isEqualTo(5);
+    }
+
+    @Test
     public void testGraalVMVersionsOlderThan() {
         assertOlderThan("GraalVM Version 19.3.6 CE", "GraalVM Version 20.2.0 (Java Version 11.0.9)");
         assertOlderThan("GraalVM Version 20.0.0 (Java Version 11.0.7)", "GraalVM Version 20.1.0 (Java Version 11.0.8)");

--- a/core/runtime/src/main/java/io/quarkus/runtime/graal/GraalVM.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/graal/GraalVM.java
@@ -72,7 +72,9 @@ public final class GraalVM {
         // Temporarily work around https://github.com/quarkusio/quarkus/issues/36246,
         // till we have a consensus on how to move forward in
         // https://github.com/quarkusio/quarkus/issues/34161
-        protected static final Map<String, String> GRAAL_MAPPING = Map.of("22", "24.0",
+        protected static final Map<String, String> GRAAL_MAPPING = Map.of(
+                "21", "23.1",
+                "22", "24.0",
                 "23", "24.1",
                 "24", "24.2",
                 "25", "25.0");

